### PR TITLE
Statelessness test: Does the CNF use a non-cloud native data store: local volumes on the node? #269

### PR DESCRIPTION
**Pull request by wavell**
# Statelessness test: Does the CNF use a non-cloud native data store: local volumes on the node? #269

## Description
Statelessness test: Does the CNF use a non-cloud native data store: local volumes on the node? #269

## Issues:
Refs: #269

## How has this been tested:
 - [ ] Covered by existing integration testing
 - [x] Added integration testing to cover
 - [ ] Verified all A/C passes
     * [ ] develop
     * [ ] master
     * [ ] tag/other branch
 - [ ] Test environment
    * [ ] Shared Packet K8s cluster
    * [ ] New Packet K8s cluster
    * [ ] Kind cluster
 - [ ] Have not tested

## Types of changes:
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
 - [ ] Documentation update

## Checklist:
**Documentation**
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] No updates required.

**Code Review**
- [ ] Does the test handle fatal exceptions, ie. rescue block

**Issue**
- [ ] Tasks in issue are checked off

### CODE REVIEW

File: sample-cnfs/sample-local-storage/chart/templates/persistent-volume.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  local:
+    path: /var/tmp
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: kubernetes.io/hostname
+          operator: In
+          values:
+          - kind-control-plane
```
Comment (taylor) [inline sample-cnfs/sample-local-storage/chart/templates/persistent-volume.yaml:22]: @denverwilliams @wavell FYI, this test will not work on a non-kind cluster (unless someone happens to name one of their nodes "kind-control-plane").

We probably need to look at another affinity item to use or add support for dynamically setting the node name.

File: sample-cnfs/sample-local-storage/chart/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+          mountPath: /home
+        resources:
+{{ toYaml .Values.resources | indent 10 }}
+        ports:
+{{ include "coredns.containerPorts" . | indent 8 }}
+      volumes:
+        - name: state-volume
+          hostPath:
+            path: /tmp/{{ randAlphaNum 10 }}
+            type: DirectoryOrCreate
```
Comment (taylor) [inline sample-cnfs/sample-local-storage/chart/templates/deployment.yaml:86]: @wavell @denverwilliams This does not mean that are actively using a local volume to retrieve data that was written there in the past.  They could still be using a local volume for temporary storage.

**Review by taylor — APPROVED** [review]
left some comments.  looks like a good first version.

my main suggestion is about using a known path every time and checking for existing data as the test vs random.  But if it's past in as a template that could be random or the same every time. Not sure how to check.

Is it okay to use a local volume if it's only for ephemeral data?

**Review by agentpoyo — APPROVED** [review]
LGTM. :+1: 

As @taylor  pointed out, my only concern would be that it's only going to work on kind cluster.

File: sample-cnfs/sample-local-storage/chart/templates/persistent-volume.yaml (hunk 1 above)
Comment (wavell) [reply sample-cnfs/sample-local-storage/chart/templates/persistent-volume.yaml]: Good catch, we addressed this today

File: sample-cnfs/sample-local-storage/chart/templates/deployment.yaml (hunk 2 above)
Comment (wavell) [reply sample-cnfs/sample-local-storage/chart/templates/deployment.yaml]: We aren't looking in the deployment yaml but in the persistent volume claim yaml.  We are looking for the existence of the local: and path: keys

Comment (denverwilliams) [reply sample-cnfs/sample-local-storage/chart/templates/deployment.yaml]: @taylor This hostPath code has nothing to do with the new test, it's a remnant from the old hostpath sample-cnf we used as a base. It's been removed it from the helm chart now.

**Pull request merged**